### ISSUE TITLE
xcode 10.1 Unexpected '@' in program  error

### ISSUE THREAD

**Issue by wapony** [SDWebImage/SDWebImage#2683]
### New Issue Checklist

if .m file have @weakify(self) or @strongify(self) , report  Unexpected '@' in program  error

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [ ] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [ ] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | e.g. ios / macos / tvos / watchos
 Platform Version        | e.g. 12.0 / 10.14.0 / 12.0 / 5.0
 SDWebImage Version      | e.g. 5.0.0 / 4.4.0
 Integration Method      | e.g. carthage / cocoapods / manually
 Xcode Version           | e.g. Xcode 10 / Xcode 9
 Repro rate              | e.g. all the time (100%) / sometimes x% / only once
 Repro with our demo prj | e.g. does it happen with our demo project?
 Demo project link       | e.g. link to a demo project that highlights the issue

### Issue Description and Steps

Please fill in the detailed description of the issue (full output of any stack trace, compiler error, ...) and the steps to reproduce the issue.

**Comment by Insofan** [comment]
Please fill the correct information. This issue maybe related to this [# 2680](https://github.com/SDWebImage/SDWebImage/issues/2680)

**Comment by dreampiggy** [comment]
@wapony Strange. Seems dunplicated of #2680.

What's your environment ? I test with Xcode 10.2 with both Cocoapods && Carthage && Manual Install (open `SDWebImage.xcodeproj` and build). They all works.

Could you please more information about this ?

**Comment by GuYueZhaoJinRen** [comment]
没有用cocoapods，也出现了这个问题，请问怎么解决

**Comment by GuYueZhaoJinRen** [comment]
没有用cocoapods，也出现了这个问题，请问怎么解决

**Comment by Insofan** [comment]
> 没有用cocoapods，也出现了这个问题，请问怎么解决

Please provide more information about your integration method or you can follow #2680 suggestion re-install sd using cocoapods.

**Comment by dreampiggy** [comment]
@GuYueZhaoJinRen How do you do using the manual install steps ?

Seems I found the wiki pages for [Manual Install Guide](https://github.com/SDWebImage/SDWebImage/blob/master/Docs/ManualInstallation.md) is not been updated. 

Actually, all you need to do, it's just as simple as you image.

## Manual Install

1. Clone the repo: `git clone https://github.com/SDWebImage/SDWebImage.git`
2. Open the `SDWebImage.xcodeproj`
3. Select the framework target you need.

+ `SDWebImage` for dynamic framework.
+ `SDWebImage Static` for static library.
+ `SDWebImageMapKit` for MapKit sub component only.

4. Select the correct platform you need.

+ `My Mac` for macOS platform.
+ `Generic iOS Device` for iOS platform.
+ `Generic tvOS Device` for tvOS platform.
+ `Generic watchOS Device` for watchOS platform.

+ Click `Archive`, then export it. There are a `SDWebImage.framework` or `libSDWebImage.a` in the build folder.

**Comment by wapony** [comment]
I found the reason why xcode reports the error.   I just copied the SDWebImage folder into my project, but the SDInternalMacros.h file did not imported default. there are many macros in this file .  I imported this file in my project's PrefixHeader.h.  Then the issue was fixed.

**Comment by dreampiggy** [comment]
> I found the reason why xcode reports the error.   I just copied the SDWebImage folder into my project, but the SDInternalMacros.h file did not imported default. there are many macros in this file .  I imported this file in my project's PrefixHeader.h.  Then the issue was fixed.

It’s not the correct way for manual install. And we can not provide any help for this. Use as your own risk.

The correct way of manual install is listed above. You can link the `libSDWebImage.a` or `SDWebImage.framework`.

If you must need the source code, using CocoaPods and don’t enable `use_framworks!`

**Comment by dreampiggy** [comment]
@wapony I update the [Manual Installation Guide](https://github.com/SDWebImage/SDWebImage/wiki/Installation-Guide#manual-installation-guide) Wiki page. It contains the both use case:

1. Build framework or Static Library as binary for Application
2. Embed `SDWebImage.xcodeproj` to use source code in your Application workspace.

You can have a see, to see whether that can help for you. It's not recommanded to copy the source files into your project. Because we may add many Xcode configuration (Such as Header Search Path, module map, Warning flags) any time, which may cause your application compile failed because you only copy the source code, without any configuration related to them.

**Comment by zhongwuzw** [comment]
@dreampiggy I remember I encounter this error before when I run our demo, but If I re-build the project, errors disappears. I suspect wether precompiled header always works.

**Comment by dreampiggy** [comment]
@zhongwuzw It seems the Xcode 10.2 + New Build System with two Pods project cause the issue. Using Xcode 10.1 does not cause two Pods project.

I change back to Legacy Build System and solve this issue...Maybe we need a better way (Such as, don't put the `Test.xcodeproj` into `SDWebImage.xcworkspace`.) or waiting for CocoaPods to solve the issue.


![image](https://user-images.githubusercontent.com/6919743/55866753-92b32b80-5bb3-11e9-8a6f-96ef2f70f915.png)

**Comment by zhongwuzw** [comment]
@dreampiggy We may have some choices:
1. Using individual `*.xcworkspace` to run tests.
2. Integrate `Tests` to the target of `SDWebImage`, only one `*.xcworkspace`.

**Comment by zhongwuzw** [comment]
@dreampiggy Is this fixed that you closed it?

I plan to revert the pch file mechanism, instead using explicit `#import`. The reasons are:
1. PCH already deprecated by Apple and disabled by default. (Detailed reason can see how Apple said.)
2. I don't think **framework** should use PCH to facilitate the development. Seems I never see any popular framework would use pch file. And even in users normal project development, they also should avoid to use pch file, cons see below.
3. CocoaPods `prefix_header_contents` is also **not recommended** .https://guides.cocoapods.org/syntax/podspec.html#prefix_header_contents

<img width="864" alt="image" src="https://user-images.githubusercontent.com/5061845/57537209-5e7c8780-7378-11e9-8b03-fd0e6861bdc1.png">


PS.
PCH cons:
1. It's controversy with Module mechanism.
2. Import pch contents in every implementation files, may bring much unnecessary code.
3. Precompile Prefix Header option sets to `NO`, not benefit for build time.
<img width="593" alt="image" src="https://user-images.githubusercontent.com/5061845/57536835-9afbb380-7377-11e9-895a-3b959b9a910c.png">

cc. @bpoplauschi @SDWebImage/collaborators .

**Comment by bpoplauschi** [comment]
I don't have detailed knowledge about the downsides of using precompiled headers (except for the extra precompiler time needed to expand those macros and duplication in all of the files)

Are you referring to any work we have previously done in this area?

**Comment by dreampiggy** [comment]
Actually, it's not so important or not to use prefix header files for simple use case.

SDWebImage use prefix header files, just because of the last time #2646 , it will introduce many many code in implementation changes. And instead of place the import again and again, use the prefix header files.

It's OK to use it or not. And does not have any pros or cons.

> It's controversy with Module mechanism.

Actually, since the prefix header files used by SD, it's `Private Prefix Header Files`, which apply only for SDWebImage itself, not cross-project. This does not cause any modular issue. This does not effect the user project or dependency project. And if you prebuilt SDWebImage as framework, all of this has been disappear.

Why prefix header files is not recommaned it's the `Public Prefix Header Files`, which will effect the user where include this project. And will impact the compile time when the pch changes. (But you can use the `Precompile Prefix Headers` to speed up)

**Comment by dreampiggy** [comment]
It's OK to change it to normal import. Just like [IGListMacros.h](https://github.com/Instagram/IGListKit/blob/master/Source/Common/IGListMacros.h) and the import file [IGListReloadDataUpdater.m](https://github.com/Instagram/IGListKit/blob/8d8a632b679b0a684e17bec2b81171da62893a9b/Source/IGListReloadDataUpdater.h#L10).

@Insofan Help for changing this ? Or I can help for change them one by one.
